## Практическое задание: pandas + matplotlib

In [216]:
!pip install pandas matplotlib gdown

Defaulting to user installation because normal site-packages is not writeable



[notice] A new release of pip available: 22.3.1 -> 24.2
[notice] To update, run: python.exe -m pip install --upgrade pip


# Часть 1. Pandas

In [217]:
import pandas as pd

In [218]:
!pip install --upgrade --no-cache-dir gdown

Defaulting to user installation because normal site-packages is not writeable



[notice] A new release of pip available: 22.3.1 -> 24.2
[notice] To update, run: python.exe -m pip install --upgrade pip


In [219]:
!gdown 1sA-fqJb1WxhS_jaYDcIwXWa_b8b2oZJ6&confirm=t

"gdown" �� ���� ����७��� ��� ���譥�
��������, �ᯮ��塞�� �ணࠬ��� ��� ������ 䠩���.
"confirm" �� ���� ����७��� ��� ���譥�
��������, �ᯮ��塞�� �ணࠬ��� ��� ������ 䠩���.


In [220]:
# data = Your code here

df = pd.read_csv('youtube_trending_mediterranean.csv', sep = ',', encoding = 'UTF-8')
df.tail(5)

country     video_id                                              title  \
3216     NaN  eB-A1Y0Dswc                             Thank you for watching   
3217     NaN  KJGTtzccOHU  Pokemon Emerald but every battle is extremely ...   
3218     NaN  be8mqaTi_DI  Victor Cibrian x Fuerza Regida x Luis R Conriq...   
3219     NaN  14aYHc7ybu0  WE STAYED AT A 1 STAR HOTEL FOR 24 HOURS | coc...   
3220     NaN  mQ9eFI13NCQ  I 100%'d Cuphead: The Delicious Last Course, H...   

               publishedAt                 channelId  \
3216  2022-07-05T21:18:33Z  UCS5tt2z_DFvG7-39J3aE-bQ   
3217  2022-07-05T01:55:10Z  UC0VVYtw21rg2cokUystu2Dw   
3218  2022-07-04T21:00:36Z  UC648rgJOboZlgcDbW00vTSA   
3219  2022-07-05T13:40:04Z  UCSGoIq_tVESqNYF1Re-zn1Q   
3220  2022-07-04T16:04:33Z  UCRgzckrm1Svf-Y_dIWsTPTg   

                  channelTitle  categoryId trending_date  \
3216             Life of Boris          22      22.11.07   
3217                  SmallAnt          20      22.11.07   
3218            Rancho Humilde          10      22.11.07   
3219                 HJ Evelyn          24      22.11.07   
3220  The Andrew Collette Show          20      22.11.07   

                                                   tags  view_count  \
3216  lifeofboris|life of boris|english subtitles|ye...      864216   
3217  pokemon|smallant|smallant1|challenge|pokemon c...     1311698   
3218  Rancho Humilde|Legado 7|El De La Guitarra|Fuer...     2256933   
3219  hjevelyn|Evelyn ha|Emily ha|erica ha|korean si...      922668   
3220  cuphead|cuphead dlc|cuphead the delicious last...     1176289   

      comment_count                                  thumbnail_link  \
3216          14431  https://i.ytimg.com/vi/eB-A1Y0Dswc/default.jpg   
3217           1232  https://i.ytimg.com/vi/KJGTtzccOHU/default.jpg   
3218            597  https://i.ytimg.com/vi/be8mqaTi_DI/default.jpg   
3219           2951  https://i.ytimg.com/vi/14aYHc7ybu0/default.jpg   
3220            620  https://i.ytimg.com/vi/mQ9eFI13NCQ/default.jpg   

      comments_disabled  ratings_disabled  \
3216              False              True   
3217              False              True   
3218              False              True   
3219              False              True   
3220              False              True   

                                            description  
3216  New channel: https://www.youtube.com/channel/U...  
3217  Pokemon emerald but every battle is a 1v2You s...  
3218  Suscríbete: https://Rancho.lnk.to/RHYoutube\r...  
3219  never again ahaha.... CHECK OUT OUR MERCH: htt...  
3220  My Cuphead: The Delicious Last Course Let's Pl...

#### Задание 1
* Откройте файл с таблицей, загруженной выше, и выведите последние 5 строк.
* Посмотрите на данные и скажите, что они из себя представляют, сколько в таблице строк и столбцов?

In [221]:
# Your code here
df.shape

(3221, 15)

**Ответ**: ВАШ ОТВЕТ ЗДЕСЬ

In [222]:
df.dtypes

country              object
video_id             object
title                object
publishedAt          object
channelId            object
channelTitle         object
categoryId            int64
trending_date        object
tags                 object
view_count            int64
comment_count         int64
thumbnail_link       object
comments_disabled      bool
ratings_disabled       bool
description          object
dtype: object

#### Задание 2 

Ответьте на вопросы:

* Для скольких стран (колонка country) есть информация о видеороликах в таблице (отсутствие страны НЕ учитывается как +1 значение)?
* Сколько есть видеороликов (== записей в таблице) c разрешенными и запрещенными комментариями? (колонка comments_disabled). Под видеороликом считаем 1 запись в таблице.
* Какое максимальное число просмотров у видеороликов, попавших в таблицу? (колонка view_count)

In [223]:
# Your code here
df.loc[df['comments_disabled'] == True]['comments_disabled'].count()

np.int64(28)

In [224]:
df.loc[df['comments_disabled'] == False]['comments_disabled'].count()

np.int64(3193)

In [225]:
df['view_count'].max()

np.int64(123034293)

**Ответ**:
1. ВАШ ОТВЕТ ЗДЕСЬ
2. Разрешено 28, запрещено 3193
3. 123034293

#### Задание 3

* Есть ли в данных пропуски? В каких колонках?

In [226]:
# Your code here
df.isna().sum()

country              200
video_id               0
title                  0
publishedAt            0
channelId              0
channelTitle           0
categoryId             0
trending_date          0
tags                   0
view_count             0
comment_count          0
thumbnail_link         0
comments_disabled      0
ratings_disabled       0
description          131
dtype: int64

In [227]:
df.dtypes

country              object
video_id             object
title                object
publishedAt          object
channelId            object
channelTitle         object
categoryId            int64
trending_date        object
tags                 object
view_count            int64
comment_count         int64
thumbnail_link       object
comments_disabled      bool
ratings_disabled       bool
description          object
dtype: object

**Ответ**: country 200, description 131

In [228]:
df.fillna(
    {
        "description": " ",
    },
    inplace=True,
)

In [229]:
df = df.dropna()

Заполните пропуски пустой строкой для текстовых колонок, и удалите строки с пропусками для колонок с категориальными значениями.

**Внимание!** Для проверки корректности выведите кол-во пропусков во всех столбцах и общее кол-во строк в таблице - должна быть 3021 запись. Ошибка при удалении строк может привести к неправильному выполнению следующих заданий.

In [230]:
df.shape

(3021, 15)

In [231]:
assert(df.shape[0] == 3021, "Проверьте правильность выполнения задания.")

<>:1: SyntaxWarning: assertion is always true, perhaps remove parentheses?
<>:1: SyntaxWarning: assertion is always true, perhaps remove parentheses?
C:\Users\Examen\AppData\Local\Temp\ipykernel_804\1825482647.py:1: SyntaxWarning: assertion is always true, perhaps remove parentheses?
  assert(df.shape[0] == 3021, "Проверьте правильность выполнения задания.")


#### Задание 4 

Посмотрите на колонку publishedAt
* Какого она типа?
* Измените тип колонки на datetime. Покажите, что после изменения колонка действительно имеет тип datetime.
* В какой месяц было добавлено больше всего видеороликов?

In [232]:
# Your code here
df['publishedAt'].dtype

dtype('O')

In [233]:
df['publishedAt'] = pd.to_datetime(df['publishedAt'])

In [234]:
df['publishedAt'].dtype

datetime64[ns, UTC]

In [235]:
grouped = df.groupby(df["publishedAt"].dt.strftime("%m")).size()

In [236]:
grouped

publishedAt
06    1066
07    1955
dtype: int64

In [237]:
x = df["publishedAt"].dt.strftime("%m") == '07'

**Ответ**:
1. О - object
2. В июле

#### Задание 5 

Посмотрите на колонку description.
* Переведите строки в нижний регистр и обрежьте ненужные пробелы справа и слева.
* Замените url-адреса  в описаниях на специальный тэг \<URL\>.
* Выведите полностью description для видеоролика с video_id = 'Oxexirlmdig'.

In [238]:
pattern = r'http\S+'

In [239]:
# Your code here
df['description'].str.lower()

0       stream or download now ''katile'' 👇 https:/...
1       the official video for gajde.stream/download g...
2       stream or download „am/pm“ here 🎧 https...
3       bzzz entertainment - https://lnk.to/bzzzstream...
4       stream „tequila“:https://taynaxazet.lnk.to...
                              ...                        
3016    fenerbahã§e hazä±rlä±k maã§ä±nda mol fehervar'...
3017    download now!! ►► https://supr.cl/2k62czgb...
3018    gã¶rã¼lmemiåŸ bir ekonomik kriz geliyor,ekonom...
3019    tozluyaka 2. bã¶lã¼m 2. fragmanä±åžans bile yo...
3020    onur tuäŸrul, mustafa demirtaåŸ, metin karabaå...
Name: description, Length: 3021, dtype: object

In [240]:
df.replace([pattern], ['<URL>'], regex=True)

country     video_id                                              title  \
0         AL  hxWSp-8jJuo                                 DON XHONI - KATILE   
1         AL  MnrC9mk2H1o  Dhurata Dora x Elvana Gjata - GAJDE (Official ...   
2         AL  5Ql-G0dTqC0                       KIDA x BUTRINT IMERI - AM/PM   
3         AL  oUdY0qSavvY                              Dafina Zeqiri - Malli   
4         AL  I4ZSvYLnOcQ                             Tayna x Azet - Tequila   
...      ...          ...                                                ...   
3016      TR  2bvhzLli-Aw   Fenerbahçe 3-0 Mol Fehervar | Hazırlık Maçı   
3017      TR  imjlTtCxkkk             Brawl Stars Season 13! - #DeepSeaBrawl   
3018      TR  rdAhWgxtteY  Görülmemiş Bir Ekonomik Kriz Geliyor- Düny...   
3019      TR  FROtCry4nLw                  Tozluyaka 2. Bölüm 2. Fragmanı   
3020      TR  cwD6jnfIi9A  FB, GS, BJK ve TS'de Son Gelişmeler | Onur & ...   

                   publishedAt                 channelId  \
0    2022-07-07 22:00:12+00:00  UCZNJXsf2I4y54M_UmMiC6Vw   
1    2022-06-30 22:00:11+00:00  UCtzhq8TIfFB9vRAHVqLL9PA   
2    2022-07-02 22:00:11+00:00  UCAHVxL9FG3Sp8ltQtLEy_oA   
3    2022-07-07 22:00:12+00:00  UCrv0PJoutcRhHkSMJYoXAIQ   
4    2022-06-30 18:00:26+00:00  UCJBZ-GcWSswmJ0xc9WAKDqw   
...                        ...                       ...   
3016 2022-07-05 18:46:34+00:00  UCpdSUUHlxMjO0c5824FGcsA   
3017 2022-07-04 09:00:09+00:00  UCooVYzDxdwTtGYAkcPmOgOw   
3018 2022-07-03 20:00:18+00:00  UC9PfAU8pdnTs7afxIXKhgVA   
3019 2022-07-01 20:47:38+00:00  UCJe13zu6MyE6Oueac41KAqg   
3020 2022-07-06 17:09:00+00:00  UCeCRRixprpBjNmzRWgLBiUA   

               channelTitle  categoryId trending_date  \
0                 DON XHONI          24      22.11.07   
1              Dhurata Dora          10      22.11.07   
2                      KIDA          10      22.11.07   
3             Dafina Zeqiri          10      22.11.07   
4     Friends Entertainment          10      22.11.07   
...                     ...         ...           ...   
3016                S Sport          17      22.11.07   
3017            Brawl Stars          20      22.11.07   
3018              Erkan Öz          25      22.11.07   
3019                    FOX          24      22.11.07   
3020                   VOLE          22      22.11.07   

                                                   tags  view_count  \
0                                                [none]     5130167   
1                                                [none]     8886648   
2     Butrint Imeri|Kida|Dale|Butrint imeri kida|But...     7230818   
3     Dafina|Zeqiri|Dafine Moj|Pa ty|Lule Lule|Duro|...     1406078   
4     tayna|Friends|Entertainment|rinor hykolli|hip ...     6397676   
...                                                 ...         ...   
3016  fenerbahçe özet|fenerbahçe|fb|fenerbahçe h...      492614   
3017  brawl stars|mobile game|mobile strategy game|m...     1062053   
3018  erkan öz yorumları|erkan öz yatırım|erkan...      115149   
3019  Tozluyaka 2. Bölüm 1. Fragmanı|emre kınay|...      901830   
3020                                             [none]      185855   

      comment_count thumbnail_link  comments_disabled  ratings_disabled  \
0              3641          <URL>              False              True   
1              4414          <URL>              False              True   
2              7518          <URL>              False              True   
3               827          <URL>              False              True   
4              6063          <URL>              False              True   
...             ...            ...                ...               ...   
3016            683          <URL>              False              True   
3017           1745          <URL>              False              True   
3018            422          <URL>              False              True   
3019            579          <URL> 

In [241]:
df.loc[df['video_id'] == 'Oxexirlmdig']['description']

12    Music line & Lyrics : GHETTO GEASY & MAJKMusic...
Name: description, dtype: object

**Ответ**:  Music line & Lyrics : GHETTO GEASY & MAJKMusic...

#### Задание 7 
* Какой видеоролик был опубликован в наибольшем кол-ве стран? В качестве ответа привeдите название видеоролика (колонка title).

Для решения могут помочь функции [groupby](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.groupby.html) и [sort_values](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sort_values.html).

In [242]:
# Your code here
df.groupby('title')['country'].count().sort_values()

title
🔴CRAZYCRAFT DOCTOR OCTOPUS YAPIYORUZ!                                                           1
🔴 LIVE | West Indies v Bangladesh | 2nd T20I                                                    1
🔥 First collective 𝚃𝚁𝙰𝙸𝙽𝙸𝙽𝙶 of the season 🔛⚽                       1
🐭 Ma ratatouille                                                                                1
🎾 Résumé - Wimbledon : Rafael Nadal – Taylor Fritz : Un match légendaire de 4 heures !     1
                                                                                                  ..
Charlie Puth - Left And Right (feat. Jung Kook of BTS) [Official Video]                           12
World's Fastest Car!                                                                              14
j-hope 'MORE' Official MV                                                                         14
God of War Ragnarök - Father and Son Cinematic Trailer | PS5 & PS4 Games            

**Ответ**: Race Highlights | 2022 British Grand Prix

#### Задание 8
* Какое максимальное число просмотров (колонка view_count) у видеороликов, выпущенных во Франции или Италии (в колонке country значениe FR или IT) в июле (колонка publishedAt)?

Необходимо придумать решение в одну строку, используя для выбора строк несколько условий, объединенных с помощью | вместо or, & вместо and и ~ вместо not, а также используя функцию [isin](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.isin.html).

In [243]:
# Your code here
df.loc[df['country'].isin(['FR', 'IT'] & df["publishedAt"].dt.strftime("%m") == '07')]['view_count'].max()

C:\Users\Examen\AppData\Local\Temp\ipykernel_804\583849290.py:2: FutureWarning: Logical ops (and, or, xor) between Pandas objects and dtype-less sequences (e.g. list, tuple) are deprecated and will raise in a future version. Wrap the object in a Series, Index, or np.array before operating instead.
  df.loc[df['country'].isin(['FR', 'IT'] & df["publishedAt"].dt.strftime("%m") == '07')]['view_count'].max()


ValueError: operands could not be broadcast together with shapes (2,) (3021,) 

**Ответ**: ВАШ ОТВЕТ ЗДЕСЬ

#### Задание 9 
* Посмотрите на названия каналов из 17 категории (колонка categoryId). Как вы думаете, что это за категория?
* Предположим, что была допущена ошибка, и все видеоролики из 17 категории, относятся к 16 категории. Исправьте нужные значения в колонке categoryId, после чего выведите все уникальные категории.

Обратите внимание, что при выборе полей с помощью оператора индексирования \[\] возвращается копия выбранных значений. При попытке присвоить таким образом новые значения для некоторых полей таблицы, будет выдано предупреждение. Более надежный способ - использовать операторы loc и iloc.

In [244]:
# Your code here
df.loc[df['categoryId'] == 17, 'categoryId'] = df.loc[df['categoryId'] == 17, 'categoryId'].replace(17, 16)


In [249]:
df['categoryId'].unique()

array([24, 10,  1, 15, 22, 23, 27, 16, 20, 26, 25,  2, 19, 28])

**Ответ**: ВАШ ОТВЕТ ЗДЕСЬ

#### Задание 10 
Посмотрите на колонку tags.

* Замените строки со значением '[none]' на пустую строку.
* Какой самый популярный тэг? Обратите внимание, что у каждого видеоролика может быть более 1 тэга.

**Подсказка 1**

Чтобы получить правильное значение, не забудьте обработать множество тэгов - убрать пробельные символы по краям и привести строки в нижний регистр.

**Подсказка 2**

Нужный тэг встречается больше 130 раз.

In [252]:
# Your code here
df.loc[df['tags'] == '[none]', 'tags'] = df.loc[df['tags'] == '[none]', 'tags'].replace('none', '')

In [254]:
df['tags']

0                                                  [none]
1                                                  [none]
2       Butrint Imeri|Kida|Dale|Butrint imeri kida|But...
3       Dafina|Zeqiri|Dafine Moj|Pa ty|Lule Lule|Duro|...
4       tayna|Friends|Entertainment|rinor hykolli|hip ...
                              ...                        
3016    fenerbahçe özet|fenerbahçe|fb|fenerbahçe h...
3017    brawl stars|mobile game|mobile strategy game|m...
3018    erkan öz yorumları|erkan öz yatırım|erkan...
3019    Tozluyaka 2. Bölüm 1. Fragmanı|emre kınay|...
3020                                               [none]
Name: tags, Length: 3021, dtype: object

**Ответ**: ВАШ ОТВЕТ ЗДЕСЬ

# Часть 2. Визуализация данных


### Оформление графиков

Важно, чтобы все графики выглядели опрятно и соответствовали следующим правилам:

- должно быть название (plt.title) графика;
- на графиках должны быть подписаны оси (plt.xlabel, plt.ylabel);
- должны быть подписаны единицы измерения (если это возможно);
- все названия должны быть понятны любому человеку, знакомому с терминологией, без заглядывания в код;
- подписи тиков на осях не должны сливаться как на одной оси, так и между ними;
- если изображено несколько сущностей на одном холсте (например несколько функций), то необходима поясняющая легенда (plt.legend);
- все линии на графиках должны быть чётко видны (нет похожих цветов или цветов, сливающихся с фоном);
- если отображена величина, имеющая очевидный диапазон значений (например, проценты могут быть от 0 до 100), то желательно масштабировать ось на весь диапазон значений (исключением является случай, когда вам необходимо показать малое отличие, которое незаметно в таких масштабах);
- графики должны быть не супер-микро и не супер-макро по размерам, так, чтобы можно было увидеть все, что нужно.
- при необходимости улучшения наглядности графиков, можно пользоваться логарифмической шкалой по осям x/y.

In [ ]:
import matplotlib
import matplotlib.pyplot as plt

#отображение графиков в ноутбуке
%matplotlib inline

#для четкой прорисовки графиков - осторожно!!! Может съедать много памяти
#закоменьте эту строку, если это случилось
%config InlineBackend.figure_format = 'svg'

### Задание 1

Постройте график (line plot) зависимости кол-ва выпущенных роликов от даты публикации.

Обратите внимание, что один и тот же ролик может быть опубликован в нескольких странах, и ему может соответствовать несколько строк в таблице. На графике необходимо учесть каждый ролик только 1 раз.

In [ ]:
# Your code here

### Задание 2
Постройте диаграмму  (bar plot) кол-ва опубликованных роликов в зависимости от страны.

In [ ]:
# Your code here

### Задание 3 
Постройте график (scatter plot) зависимости кол-ва комментариев для ролика от кол-ва просмотров для роликов во Франции (country='FR'). На графике не нужно учитывать ролики с отключенными комментариями (comments_disabled=True). Какой характер зависмости получился?

In [ ]:
# Your code here

**Ответ**: ВАШ ОТВЕТ ЗДЕСЬ